File: src/float80.c
#include <math.h>
#include <stdio.h>
#include <string.h>

#include "erl_nif.h"

const int FLOAT80_SIZE = 10;

/**
 * See the website for a description of how to write a NIF:
 * https://andrealeopardi.com/posts/using-c-from-elixir-with-nifs/
 * 
 * Terms:
 * - Float80: An Elixir value represented using a 10-byte binary
 * - long double: A C type using an 80-bit extended precision double
 */

/**
 * Debug function to print bytes in Float80 binary.
 */
static void printBytes(char *label, ErlNifBinary *bin) {
    printf("%s: \n", label);
    printf("\tsize: %ld bytes\n", bin->size);

    printf("\t<<");
    for (int i = 0; i < bin->size; i++) {
        printf("%d,", bin->data[i]);
    }
    printf(">>\n");
}

/**
 * Convenience function used by most NIF functions to return
 * a Float80 value.
 */
static ERL_NIF_TERM
long_double_to_term(ErlNifEnv *env, long double value) {
    ERL_NIF_TERM term;
    unsigned char *data = enif_make_new_binary(env, FLOAT80_SIZE, &term);
    memcpy(data, &value, FLOAT80_SIZE);
    return term;
}

/**
 * Converts a binary term to a long double.
 */
static long double
term_to_long_double(ErlNifEnv *env, ERL_NIF_TERM term) {

    long double result = 0;
    ErlNifBinary bin;

    if (enif_term_to_binary(env, term, &bin)) {

        if (bin.size >= FLOAT80_SIZE) {
            // The size is 16 with the 10 byte data right aligned.
            // So increase pointer by the difference.
            // Should take into account endianness?
            unsigned char *data = bin.data + (bin.size - FLOAT80_SIZE);

            long double *value = (long double *) data;
            result = *value;
        }

        enif_release_binary(&bin);
    }

    return result;
}

/**
 * Takes an Erlang float and returns a binary Float80.
 */
static ERL_NIF_TERM
from_float(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
    double a = 0;
    enif_get_double(env, argv[0], &a);
    long double result = a;
    return long_double_to_term(env, result);
}

/**
 * Takes an Erlang integer and returns a binary Float80.
 */
static ERL_NIF_TERM
from_integer(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
    long a = 0;
    enif_get_int64(env, argv[0], &a);
    long double result = a;
    return long_double_to_term(env, result);
}

/**
 * Takes a binary Float80 and returns an Erlang float.
 * Note that this will result in a loss of prevision.
 */
static ERL_NIF_TERM
to_float(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
    long double a = term_to_long_double(env, argv[0]);
    double result = a;
    return enif_make_double(env, result);
}

/**
 * Takes two Float80 arguments and returns the sum as a Float80.
 */
static ERL_NIF_TERM
add(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);
    long double b = term_to_long_double(env, argv[1]);

    long double result = a + b;

    return long_double_to_term(env, result);
}

/**
 * Takes two Float80 arguments and returns the difference (a-b) as a Float80.
 */
static ERL_NIF_TERM
subtract(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);
    long double b = term_to_long_double(env, argv[1]);

    long double result = a - b;

    return long_double_to_term(env, result);
}

/**
 * Takes two Float80 arguments and returns the product as a Float80.
 */
static ERL_NIF_TERM
multiply(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);
    long double b = term_to_long_double(env, argv[1]);

    long double result = a * b;

    return long_double_to_term(env, result);
}

/**
 * Takes two Float80 arguments and returns the division (a/b) as a Float80.
 */
static ERL_NIF_TERM
divide(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);
    long double b = term_to_long_double(env, argv[1]);

    long double result = a / b;

    return long_double_to_term(env, result);
}

/**
 * Absolute value of Float80, returns Float80 result.
 */
static ERL_NIF_TERM
abs80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
    long double a = term_to_long_double(env, argv[0]);
    long double result = fabsl(a);
    return long_double_to_term(env, result);
}

/**
 * PI returned as Float80.
 */
static ERL_NIF_TERM
pi(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
    long double result = M_PI;
    return long_double_to_term(env, result);
}

/**
 * Arcosine of Float80, returns Float80 result.
 */
static ERL_NIF_TERM
acos80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = acosl(a);

    return long_double_to_term(env, result);
}

/**
 * Arcsine of Float80, returns Float80 result.
 */
static ERL_NIF_TERM
asin80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = asinl(a);

    return long_double_to_term(env, result);
}

/**
 * Arctangent of Float80, returns Float80 result.
 */
static ERL_NIF_TERM
atan80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = atanl(a);

    return long_double_to_term(env, result);
}

/**
 * Cosine of Float80 (in radians), returns Float80 result.
 */
static ERL_NIF_TERM
cos80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = cosl(a);

    return long_double_to_term(env, result);
}

/**
 * Sine of Float80 (in radians), returns Float80 result.
 */
static ERL_NIF_TERM
sin80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = sinl(a);

    return long_double_to_term(env, result);
}

/**
 * Tangent of Float80, returns Float80 result.
 */
static ERL_NIF_TERM
tan80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = tanl(a);

    return long_double_to_term(env, result);
}

/**
 * e to the xth power, returns Float80 result.
 */
static ERL_NIF_TERM
exp80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = expl(a);

    return long_double_to_term(env, result);
}

/**
 * Natural logarithm, returns Float80 result.
 */
static ERL_NIF_TERM
log80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = logl(a);

    return long_double_to_term(env, result);
}

/**
 * Base 10 logarithm, returns Float80 result.
 */
static ERL_NIF_TERM
log1080(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = log10l(a);

    return long_double_to_term(env, result);
}

/**
 * Base 2 logarithm, returns Float80 result.
 */
static ERL_NIF_TERM
log280(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = log2l(a);

    return long_double_to_term(env, result);
}

/**
 * a to the power of b, returns Float80 result.
 */
static ERL_NIF_TERM
pow80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);
    long double b = term_to_long_double(env, argv[1]);

    long double result = powl(a, b);

    return long_double_to_term(env, result);
}

/**
 * Square root, returns Float80 result.
 */
static ERL_NIF_TERM
sqrt80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = sqrtl(a);

    return long_double_to_term(env, result);
}

/**
 * Round up, returns Float80 result.
 */
static ERL_NIF_TERM
ceil80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = ceill(a);

    return long_double_to_term(env, result);
}

/**
 * Round down, returns Float80 result.
 */
static ERL_NIF_TERM
floor80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = floorl(a);

    return long_double_to_term(env, result);
}

/**
 * Round to nearest, returns Float80 result.
 */
static ERL_NIF_TERM
round80(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {

    long double a = term_to_long_double(env, argv[0]);

    long double result = roundl(a);

    return long_double_to_term(env, result);
}

// Let's define the array of ErlNifFunc beforehand:
static ErlNifFunc nif_funcs[] = {
    // {erl_function_name, erl_function_arity, c_function}
    {"from_float", 1, from_float},
    {"from_integer", 1, from_integer},
    {"to_float", 1, to_float},
    // Arithmetic
    {"add", 2, add},
    {"subtract", 2, subtract},
    {"multiply", 2, multiply},
    {"divide", 2, divide},
    {"abs", 1, abs80},
    // Constants
    {"pi", 0, pi},
    // Trigonometry
    {"acos", 1, acos80},
    {"asin", 1, asin80},
    {"atan", 1, atan80},
    {"cos", 1, cos80},
    {"sin", 1, sin80},
    {"tan", 1, tan80},
    // Powers
    {"exp", 1, exp80},
    {"log", 1, log80},
    {"log10", 1, log1080},
    {"log2", 1, log280},
    {"pow", 2, pow80},
    {"sqrt", 1, sqrt80},
    // Rounding
    {"ceil", 1, ceil80},
    {"floor", 1, floor80},
    {"round", 1, round80},
};

ERL_NIF_INIT(Elixir.Float80, nif_funcs, NULL, NULL, NULL, NULL)

File: lib/float80.ex
defmodule Float80 do
  @moduledoc """
  Documentation for `Float80`.

  The Float80 module allows 80 bit precision floating point numbers to be
  represented as 10 byte binary values (`<<_::80>>`). NIFs are used to convert Elixir integers 
  and floats to and from Float80 and also perform math functions returning the result
  as a Float80.

  Note that since Elixir doesn't support Float80 numbers you will need to convert
  Float80 numbers back to float before passing them to other Elixir functions.

  The standard arithmetic operators are not supported (`+`, `-`, `*`, `/`) instead use
  the functions: `Float80.add/2`, `Float80.subtract/2`, `Float80.multiple/2`, `Float80.divide/2`.

  ## Examples
      iex> a = Float80.from(1.0)
      <<0, 0, 0, 0, 0, 0, 0, 128, 255, 63>>
      iex> b = Float80.from(2.0)
      <<0, 0, 0, 0, 0, 0, 0, 128, 0, 64>>
      iex> result80 = Float80.add(a, b)
      <<0, 0, 0, 0, 0, 0, 0, 192, 0, 64>>
      iex> result = Float80.to_float(result80)
      3.0

  """

  @typedoc """
  80-bit floating point number.
  """
  @type float80 :: <<_::80>>

  @on_load :load_nifs

  app = Mix.Project.config[:app]

  def load_nifs do
    path = :filename.join(:code.priv_dir(unquote(app)), 'float80')
    :ok = :erlang.load_nif(path, 0)
  end

  @doc """
  Returns a Float80 with the value 0.
  """
  @spec zero() :: float80
  def zero do 
    <<0::80>>
  end

  @spec from(number) :: float80
  def from(value) when is_number(value) do
    cond do
      is_integer(value) ->
        from_integer(value)
      is_float(value) ->
        from_float(value)
    end
  end

  @spec from_float(float) :: float80
  defp from_float(_) do
    exit(:nif_library_not_loaded)
  end

  @spec from_integer(integer) :: float80
  defp from_integer(_) do
    exit(:nif_library_not_loaded)
  end

  @spec to_float(float80) :: float
  def to_float(_) do
    exit(:nif_library_not_loaded)
  end

  @spec add(float80, float80) :: float80
  def add(_, _) do
    exit(:nif_library_not_loaded)
  end

  @spec subtract(float80, float80) :: float80
  def subtract(_, _) do
    exit(:nif_library_not_loaded)
  end

  @spec multiply(float80, float80) :: float80
  def multiply(_, _) do
    exit(:nif_library_not_loaded)
  end

  @spec divide(float80, float80) :: float80
  def divide(_, _) do
    exit(:nif_library_not_loaded)
  end

  @spec abs(float80) :: float80
  def abs(_) do
    exit(:nif_library_not_loaded)
  end

  @spec pi() :: float80
  def pi() do
    exit(:nif_library_not_loaded)
  end

  @spec acos(float80) :: float80
  def acos(_) do
    exit(:nif_library_not_loaded)
  end

  @spec asin(float80) :: float80
  def asin(_) do
    exit(:nif_library_not_loaded)
  end

  @spec atan(float80) :: float80
  def atan(_) do
    exit(:nif_library_not_loaded)
  end

  @spec cos(float80) :: float80
  def cos(_) do
    exit(:nif_library_not_loaded)
  end

  @spec sin(float80) :: float80
  def sin(_) do
    exit(:nif_library_not_loaded)
  end

  @spec tan(float80) :: float80
  def tan(_) do
    exit(:nif_library_not_loaded)
  end

  @spec exp(float80) :: float80
  def exp(_) do
    exit(:nif_library_not_loaded)
  end

  @spec log(float80) :: float80
  def log(_) do
    exit(:nif_library_not_loaded)
  end

  @spec log10(float80) :: float80
  def log10(_) do
    exit(:nif_library_not_loaded)
  end

  @spec log2(float80) :: float80
  def log2(_) do
    exit(:nif_library_not_loaded)
  end

  @spec pow(float80, float80) :: float80
  def pow(_, _) do
    exit(:nif_library_not_loaded)
  end

  @spec sqrt(float80) :: float80
  def sqrt(_) do
    exit(:nif_library_not_loaded)
  end

  @spec ceil(float80) :: float80
  def ceil(_) do
    exit(:nif_library_not_loaded)
  end

  @spec floor(float80) :: float80
  def floor(_) do
    exit(:nif_library_not_loaded)
  end

  @spec round(float80) :: float80
  def round(_) do
    exit(:nif_library_not_loaded)
  end

end
